### PR TITLE
AGS 4: Script API: convert Walk-areas and Walkbehinds API to OO-style

### PR DESCRIPTION
**Pull request by ivan-mogilko**
Converting walkable areas and walk-behinds script api to OO-style, in compliance with existing Hotspot, Regions etc api.

This adds two managed structs, and array properties in the Room struct for getting these.

```
builtin managed struct WalkableArea {
  /// Gets the walkable area at the specified position on the screen.
  import static WalkableArea* GetAtScreenXY(int x, int y);
  /// Returns the walkable area at the specified position within this room.
  import static WalkableArea* GetAtRoomXY(int x, int y);
  /// Gets the drawing surface for the 8-bit walkable mask
  import static DrawingSurface* GetDrawingSurface();
  /// Changes this walkable area's scaling level.
  import void SetScaling(int min, int max);
  /// Gets the ID number for this area.
  import readonly attribute int ID;
  /// Gets/sets whether this walkable area is enabled.
  import attribute bool Enabled;
  /// Gets/sets this walkable area's minimal scaling level.
  import readonly attribute int ScalingMin;
  /// Gets/sets this walkable area's maximal scaling level.
  import readonly attribute int ScalingMax;
};

builtin managed struct Walkbehind {
  /// Gets the walk-behind at the specified position on the screen.
  import static WalkableArea* GetAtScreenXY(int x, int y);
  /// Returns the walk-behind at the specified position within this room.
  import static WalkableArea* GetAtRoomXY(int x, int y);
  /// Gets the drawing surface for the 8-bit walk-behind mask
  import static DrawingSurface* GetDrawingSurface();
  /// Gets the ID number for this walk-behind.
  import readonly attribute int ID;
  /// Gets/sets this walk-behind's baseline.
  import attribute int Baseline;
};
```

Added 2 global arrays of pointers to match existing ones, since they are still in use:
```
import readonly WalkableArea *walkarea[AGS_MAX_WALKABLE_AREAS];
import readonly Walkbehind *walkbehind[AGS_MAX_WALKBEHINDS];
```

But also added Room properties that return room's elements, as a suggestion for future use:
```
  /// Accesses the Hotspots in the current room.
  import static readonly attribute Hotspot *Hotspots[];
  /// Accesses the Objects in the current room.
  import static readonly attribute Object *Objects[];
  /// Accesses the Regions in the current room.
  import static readonly attribute Region *Regions[];
  /// Accesses the Walkable areas in the current room.
  import static readonly attribute WalkableArea *WalkableAreas[];
  /// Accesses the Walk-behinds in the current room.
  import static readonly attribute Walkbehind *Walkbehinds[];
```

### CODE REVIEW

**Comment by ericoporto** [comment]
Hey, the build failures are because the new files need to be added to `Engine/CMakeLists.txt`. Nit sure if I am missing this somewhere, but the Room struct in the Script API could also have an ObjectsCount property - perhaps the same to others, or not if they are a fixed number.

**Comment by ivan-mogilko** [comment]
> Not sure if I am missing this somewhere, but the Room struct in the Script API could also have an ObjectsCount property

There's is one already.

**Comment by ivan-mogilko** [comment]
I added two global pointer arrays in addition to existing ones, since these are still in use:
```
import readonly WalkableArea *walkarea[AGS_MAX_WALKABLE_AREAS];
import readonly Walkbehind *walkbehind[AGS_MAX_WALKBEHINDS];
```

But also room properties for getting everything, as an alternative:
```
  /// Accesses the Hotspots in the current room.
  import static readonly attribute Hotspot *Hotspots[];
  /// Accesses the Objects in the current room.
  import static readonly attribute Object *Objects[];
  /// Accesses the Regions in the current room.
  import static readonly attribute Region *Regions[];
  /// Accesses the Walkable areas in the current room.
  import static readonly attribute WalkableArea *WalkableAreas[];
  /// Accesses the Walk-behinds in the current room.
  import static readonly attribute Walkbehind *Walkbehinds[];
```

File: Editor/AGS.Editor/Resources/agsdefns.sh [hunk 1]
```diff
@@ -1609,6 +1620,42 @@ builtin managed struct Region {
...
+  import static WalkableArea* GetAtScreenXY(int x, int y);    // $AUTOCOMPLETESTATICONLY$
+  /// Returns the walkable area at the specified position within this room.
+  import static WalkableArea* GetAtRoomXY(int x, int y);    // $AUTOCOMPLETESTATICONLY$
+  /// Gets the drawing surface for the 8-bit walkable mask
+  import static DrawingSurface* GetDrawingSurface();  // $AUTOCOMPLETESTATICONLY$
+  /// Changes this walkable area's scaling level.
+  import void SetScaling(int min, int max);
+  /// Gets the ID number for this area.
+  import readonly attribute int ID;
+  /// Gets/sets whether this walkable area is enabled.
```
Comment (ericoporto) [inline Editor/AGS.Editor/Resources/agsdefns.sh:1647]: `Gets/sets` in the comment should just be Gets for `ScalingMin` and `ScalingMax`

File: Engine/ac/dynobj/cc_walkablearea.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#include "game/roomstruct.h"
+#include "util/stream.h"
+
+using namespace AGS::Common;
+
+extern ScriptWalkableArea scrWalkarea[MAX_WALK_AREAS + 1];
+
+// return the type name of the object
+const char *CCWalkableArea::GetType() {
+    return "Walkarea";
```
Comment (ericoporto) [inline Engine/ac/dynobj/cc_walkablearea.cpp:27]: Should this be `WalkableArea` ?

File: Editor/AGS.Editor/Resources/agsdefns.sh [hunk 3]
```diff
@@ -1611,6 +1624,42 @@ builtin managed struct Region {
...
+  import readonly attribute int ScalingMin;
+  /// Gets this walkable area's maximal scaling level.
+  import readonly attribute int ScalingMax;
+};
+#endif
+
+#ifdef SCRIPT_API_v400
+builtin managed struct Walkbehind {
+  /// Gets the walk-behind at the specified position on the screen.
+  import static WalkableArea* GetAtScreenXY(int x, int y);    // $AUTOCOMPLETESTATICONLY$
```
Comment (ericoporto) [inline Editor/AGS.Editor/Resources/agsdefns.sh:1663]: This should be a Walkbehind pointer

**Pull request merged**